shell/packages/sandstorm-ui-powerbox/powerbox-client.js

```diff
@@ -0,0 +1,429 @@
+      grainId: grain._id,
+      title: grain.title,
+      appTitle: appTitle,
+      iconSrc: iconSrc,
+      lastUsed: grain.lastUsed,
+    };
+    cardData.callback = function () {
+      // Because Blaze always invokes functions when referenced as values from the data context, we
+      // need to double-wrap this callback.
+      return function () {
+        selectGrain(cardData);
+      };
+    };
+
+    return cardData;
+  });
+};
+
+const matchesAppOrGrainTitle = function (needle, cardData) {
+  if (cardData.title && cardData.title.toLowerCase().indexOf(needle) !== -1) return true;
+  if (cardData.appTitle && cardData.appTitle.toLowerCase().indexOf(needle) !== -1) return true;
+  return false;
+};
+
+const compileMatchFilter = function (searchString) {
+  // split up searchString into an array of regexes, use them to match against item
+  const searchKeys = searchString.toLowerCase()
+      .split(" ")
+      .filter(function (k) { return k !== "";});
+
+  return function matchFilter(item) {
+    if (searchKeys.length === 0) return true;
+    return _.chain(searchKeys)
+        .map(function (searchKey) { return matchesAppOrGrainTitle(searchKey, item); })
+        .reduce(function (a, b) { return a && b; })
+        .value();
+  };
+};
+
+const indexElements = function (arr) {
+  // Helper function to annotate an array of objects with their indices
+  for (let i = 0; i < arr.length; i++) {
+    arr[i].index = i;
+  }
+};
+
+const removeObsolete = function (arr) {
+  // remove entries from the list that are flagged as obsolete
+  return _.filter(arr, function (el) {
+    return !el.obsolete;
+  });
+};
+
+Template.powerboxRequest.onRendered(function () {
+  const searchbar = this.findAll(".search-bar")[0];
+  if (searchbar) searchbar.focus();
+});
+
+Template.powerboxRequest.helpers({
+  grains: function () {
+    const ref = Template.instance().data.get();
+    return ref.filteredCardData();
+  },
+
+  selectedProvider: function () {
+    const ref = Template.instance().data.get();
+    return ref._selectedProvider.get();
+  },
+
+  selectedProviderTemplate: function () {
+    const ref = Template.instance().data.get();
+    return ref._selectedProvider.get().templateName;
+  },
+
+  selectedProviderTemplateData: function () {
+    const ref = Template.instance().data.get();
+    return ref._selectedProvider.get().templateData();
+  },
+
+  requestedInterfaceIsImplementedByFrontendRef: function () {
+    const ref = Template.instance().data.get();
+    return (
+      // TODO: support additional frontendref types
+      //ref.requestedInterfaceMatchesTag({id: "12214421258504904768"}) || // IpNetwork
+      //ref.requestedInterfaceMatchesTag({id: "16369547182874744570"}) || // IpInterface
+      ref.requestedInterfaceMatchesTag({ id: "10145426045091321204" }) // SealedUiView
+    );
+  },
+
+  showWebkeyInput: function () {
+    // Transitional feature: treat requests that specified no query patterns to match, not even the
```

Comment:
I feel like there should always be some way to input a webkey, possibly hidden behind a click on "advanced" or some such.